arangod/Aql/ModificationExecutor.cpp

```diff
@@ -157,53 +154,89 @@ void ModificationExecutor<FetcherType, ModifierType>::doOutput(OutputAqlItemRow&
+    typename FetcherType::DataRange& input, OutputAqlItemRow& output)  // FetcherType::DataRange
+    -> std::tuple<ExecutorState, ModificationStats, AqlCall> {
+  TRI_ASSERT(_infos._trx);
 
-  // if we returned "WAITING" the last time we still have
-  // documents in the accumulator that we have not submitted
-  // yet
-  if (_lastState != ExecutionState::WAITING) {
-    _modifier.reset();
+  auto stats = ModificationStats{};
+
+  _modifier.reset();
+
+  // StateType (empty || HASMORE )
+  // (if constexp )type of range = input matrx
+  // doCollect (input.nextDataRange)
+  //   -> state
+  // else
+  // doCollect input
+  //   -> state
+
+  ExecutorState upstreamState = ExecutorState::HASMORE;
+  // only produce at most output.numRowsLeft() many results
+  if constexpr (std::is_same_v<typename FetcherType::DataRange, AqlItemBlockInputMatrix>) {
+    auto range = input.getNextInputRange();
+    doCollect(range, output.numRowsLeft());
+    upstreamState = range.upstreamState();
+  } else {
+    doCollect(input, output.numRowsLeft());
+    upstreamState = input.upstreamState();
   }
 
-  TRI_IF_FAILURE("ModificationBlock::getSome") {
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-  }
+  if (_modifier.nrOfOperations() > 0) {
+    _modifier.transact();
 
-  std::tie(_lastState, stats) = doCollect(maxOutputs);
+    if (_infos._doCount) {
+      stats.addWritesExecuted(_modifier.nrOfWritesExecuted());
+      stats.addWritesIgnored(_modifier.nrOfWritesIgnored());
+    }
 
-  if (_lastState == ExecutionState::WAITING) {
-    return {ExecutionState::WAITING, std::move(stats)};
+    doOutput(output, stats);
   }
 
-  TRI_ASSERT(_lastState == ExecutionState::DONE || _lastState == ExecutionState::HASMORE);
+  return {upstreamState, stats, AqlCall{}};
+}
 
-  _modifier.transact();
+template <typename FetcherType, typename ModifierType>
+[[nodiscard]] auto ModificationExecutor<FetcherType, ModifierType>::skipRowsRange(
+    typename FetcherType::DataRange& input, AqlCall& call)
+    -> std::tuple<ExecutorState, Stats, size_t, AqlCall> {
+  auto stats = ModificationStats{};
+  _modifier.reset();
+
+  ExecutorState upstreamState = ExecutorState::HASMORE;
+  // only produce at most output.numRowsLeft() many results
+  if constexpr (std::is_same_v<typename FetcherType::DataRange, AqlItemBlockInputMatrix>) {
+    auto range = input.getNextInputRange();
+    doCollect(range, call.getOffset());
+    upstreamState = range.upstreamState();
+  } else {
+    doCollect(input, call.getOffset());
+    upstreamState = input.upstreamState();
+  }
+
+  if (_modifier.nrOfOperations() > 0) {
+    _modifier.transact();
 
-  // If the query is silent, there is no way to relate
-  // the results slice contents and the submitted documents
-  // If the query is *not* silent, we should get one result
-  // for every document.
-  // Yes. Really.
-  TRI_ASSERT(_infos._options.silent || _modifier.nrOfDocuments() == _modifier.nrOfResults());
+    if (_infos._doCount) {
```

Comment:
Why is that not included in doOutput anymore?

Reply:
Because we need it in the skip as well.
The doOutput does not need the Stats any more to make it clearer